### PR TITLE
Make ExprVisitor into C++ class

### PR DESCRIPTION
**Pull request by binji**
Add ExprVisitor::Delegate for callbacks, and ExprVisitor::DelegateNop
when some functions aren't implemented (similar to BinaryReader).

### CODE REVIEW

File: src/apply-names.cc [hunk 1]
```diff
@@ -32,9 +33,26 @@ namespace wabt {
 
 namespace {
 
-struct Context {
+struct Context : ExprVisitor::DelegateNop {
```
Comment (sbc100) [inline src/apply-names.cc]: My gut tells me we should be using class here since there are many (virtual) methods, but maybe my gut is wrong.

File: src/generate-names.cc [hunk 2]
```diff
@@ -33,8 +34,12 @@ namespace wabt {
 
 namespace {
 
-struct Context {
-  Context() : module(nullptr), label_count(0) { WABT_ZERO_MEMORY(visitor); }
+struct Context : public ExprVisitor::DelegateNop {
+  Context() : module(nullptr), visitor(this), label_count(0) {}
```
Comment (sbc100) [inline src/generate-names.cc]: Ditto, struct inheriting from classes seems a string.

File: src/apply-names.cc (hunk 1 above)
Comment (binji) [inline src/apply-names.cc]: Agreed, just thought it might be nicer to do that in another PR
